File: year2/semester2/mpp/c#/TripClientServerGrpc/GrpcTripsServer/Net/TripServicesGrpcImpl.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Threading.Tasks.Dataflow;
using System.Xml.Linq;
using Common.Net;
using Grpc.Core;
using log4net;
using TripCommon.Validator;
using tripsGrpc;
using Account = TripCommon.domain.Account;
using Trip = TripCommon.domain.Trip;
using Booking = TripCommon.domain.Booking;

namespace GrpcTripsServer.Net
{
    class TripServicesGrpcImpl : TripServices.TripServicesBase
    {
        private static readonly ILog Logger = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);

        private ITripServices tripServices;
        private ConcurrentDictionary<int, IServerStreamWriter<Response>> observers = new ConcurrentDictionary<int, IServerStreamWriter<Response>>();
        private readonly BufferBlock<Response> buffer = new BufferBlock<Response>();

        public TripServicesGrpcImpl(ITripServices tripServices)
        {
            this.tripServices = tripServices;
        }

        public override async Task Register(Request request, IServerStreamWriter<Response> responseStream, ServerCallContext context)
        {
            Logger.InfoFormat("adding observer {0}", responseStream);
            observers[request.Account.Id] = responseStream;

            while (observers.ContainsKey(request.Account.Id))
            {
                var e = await buffer.ReceiveAsync();
                foreach (var serverStreamWriter in observers.Values)
                {
                    Logger.InfoFormat("notifying {0}", serverStreamWriter);
                    await serverStreamWriter.WriteAsync(e);
                }
            }
        }

        public override Task<Response> Logout(Request request, ServerCallContext context)
        {
            tripsGrpc.Account account = request.Account;
            IServerStreamWriter<Response> o = null;
            observers.Remove(account.Id, out o);
            Logger.InfoFormat("removing observer {0}", account.Id);
            return Task.FromResult(ProtoUtils.CreateOkResponse());
        }

        public override Task<Response> GetAccountByNamePassword(Request request, ServerCallContext context)
        {
            Account a = ProtoUtils.GetAccount(request);
            Logger.InfoFormat("getting account by name {0} and password {1}", a.Name, a.Password);
            Response response = null;
            try
            {
                Account account = tripServices.GetAccountByNamePassword(a.Name, a.Password);
                response = ProtoUtils.CreateLoginResponse(account);
            }
            catch (Exception e)
            {
                response = ProtoUtils.CreateErrorResponse(e);
            }
            Logger.InfoFormat("sending response {0}", response);
            return Task.FromResult(response);
        }

        public override Task<Response> GetAllTrips(Request request, ServerCallContext context)
        {
            Logger.Info("getting all trips");
            Response response = null;
            try
            {
                List<Trip> t = tripServices.GetAllTrips();
                response = ProtoUtils.CreateGetTripsResponse(t);
            }
            catch (Exception e)
            {
                response = ProtoUtils.CreateErrorResponse(e);
            }
            Logger.InfoFormat("sending get all trips response {0}", response.Type);
            return Task.FromResult(response);
        }

        public override Task<Response> GetTripsByLandmarkDepartureHour(Request request, ServerCallContext context)
        {
            Logger.Info("searching trips");
            Response response = null;
            try
            {
                tripsGrpc.Trip t = request.Trip;
                List<Trip> tripList = tripServices.GetTripsByLandmarkDepartureHour(t.Landmark, t.Start, t.End);
                response = ProtoUtils.CreateGetTripsResponse(tripList);
            }
            catch (Exception e)
            {
                response = ProtoUtils.CreateErrorResponse(e);
            }
            Logger.InfoFormat("sending search trips response {0}", response.Type);
            return Task.FromResult(response);
        }

        public override Task<Response> AddBooking(Request request, ServerCallContext context)
        {
            Logger.Info("add booking");
            Response response = null;
            try
            {
                tripsGrpc.Booking booking = request.Booking;
                Trip trip = ProtoUtils.ConvertToTrip(booking.Trip);
                Account account = ProtoUtils.ConvertToAccount(booking.Account);
                Booking savedBooking = tripServices.AddBooking(booking.Client, booking.Phone, booking.NumTickets, trip, account);
                response = ProtoUtils.CreateOkResponse();

                Logger.Info("notifying observers");
                Response responseBookingAdded = new Response()
                {
                    Booking = new tripsGrpc.Booking()
                    {
                        Id = savedBooking.Id,
                        Account = booking.Account,
                        Trip = booking.Trip,
                        Client = savedBooking.ClientName,
                        Phone = savedBooking.PhoneNumber,
                        NumTickets = savedBooking.NumTickets
                    }
                };
                buffer.Post(responseBookingAdded);
            }
            catch (Exception e)
            {
                response = ProtoUtils.CreateErrorResponse(e);
            }
            Logger.InfoFormat("add booking response {0}", response);
            return Task.FromResult(response);
        }
    }
}

File: year2/semester2/mpp/java/TripClientServerGrpc/TripServer/src/main/java/net/TripServicesGrpcImpl.java
package net;

import domain.Account;
import domain.Trip;
import io.grpc.stub.StreamObserver;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import service.TripServices;
import tripsGrpc.Booking;
import tripsGrpc.Request;
import tripsGrpc.Response;
import tripsGrpc.TripServicesGrpc;

import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class TripServicesGrpcImpl extends TripServicesGrpc.TripServicesImplBase {
    private static final Logger LOGGER = LogManager.getLogger(TripServicesGrpcImpl.class.getName());
    private TripServices tripServices;

    public TripServicesGrpcImpl(TripServices tripServices) {
        this.tripServices = tripServices;
    }

    private ConcurrentHashMap<Integer, StreamObserver<Response>> observers = new ConcurrentHashMap<>();

    @Override
    public void getAccountByNamePassword(Request request, StreamObserver<Response> responseObserver) {
        Account a = ProtoUtilsServer.getAccount(request);
        LOGGER.info("getting account by name {} and password {}", a.getName(), a.getPassword());
        Response response = null;
        try {
            Account account = tripServices.getAccountByNamePassword(a.getName(), a.getPassword());
            response = ProtoUtilsServer.createLoginResponse(account);
        } catch (Exception e) {
            response = ProtoUtilsServer.createErrorResponse(e);
        }
        LOGGER.info("sending response {}", response);
        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }

    @Override
    public void register(Request request, StreamObserver<Response> responseObserver) {
        LOGGER.info("adding observer " + responseObserver);
        tripsGrpc.Account account = request.getAccount();
        observers.put(account.getId(), responseObserver);
    }


    @Override
    public void getAllTrips(Request request, StreamObserver<Response> responseObserver) {
        LOGGER.info("getting all trips");
        Response response = null;
        try {
            List<Trip> t = tripServices.getAllTrips();
            response = ProtoUtilsServer.createGetTripsResponse(t);
        } catch (Exception e) {
            response = ProtoUtilsServer.createErrorResponse(e);
        }
        LOGGER.info("sending get all trips response {}", response.getType());
        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }

    @Override
    public void addBooking(Request request, StreamObserver<Response> responseObserver) {
        LOGGER.info("add booking");
        Response response = null;
        try {
            Booking booking = request.getBooking();
            Trip trip = ProtoUtils.convertToTrip(booking.getTrip());
            Account account = ProtoUtils.convertToAccount(booking.getAccount());
            domain.Booking savedBooking = tripServices.addBooking(booking.getClient(), booking.getPhone(), booking.getNumTickets(), trip, account);
            response = ProtoUtilsServer.createOkResponse();

            LOGGER.info("notifying observers");
            observers.values().forEach(o -> {
                LOGGER.info("notify " + o);
                Booking bookingDto = Booking.newBuilder().setId(savedBooking.getId())
                        .setClient(savedBooking.getClientName()).setAccount(booking.getAccount()).setTrip(booking.getTrip())
                        .setPhone(booking.getPhone()).setNumTickets(booking.getNumTickets()).build();
                Response response1 = Response.newBuilder().setBooking(bookingDto).build();
                o.onNext(response1);
            });
        } catch (Exception e) {
            response = Response.newBuilder().setType(Response.ResponseType.ERROR).setError(e.getMessage()).build();
        }
        LOGGER.info("add booking response {}", response);
        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }

    @Override
    public void getTripsByLandmarkDepartureHour(Request request, StreamObserver<Response> responseObserver) {
        LOGGER.info("searching trips");
        Response response = null;
        try {
            tripsGrpc.Trip t = request.getTrip();
            List<Trip> tripList = tripServices.getTripsByLandmarkDepartureHour(t.getLandmark(), t.getStart(), t.getEnd());
            response = ProtoUtilsServer.createGetTripsResponse(tripList);
        } catch (Exception e) {
            response = Response.newBuilder().setType(Response.ResponseType.ERROR).setError(e.getMessage()).build();
        }
        LOGGER.info("sending search trips response {}", response.getType());
        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }

    @Override
    public void logout(Request request, StreamObserver<Response> responseObserver) {
        tripsGrpc.Account account = request.getAccount();
        observers.remove(account.getId());
        LOGGER.info("removing observer {}",  account.getId());
        responseObserver.onNext(ProtoUtilsServer.createOkResponse());
        responseObserver.onCompleted();
    }
}
//    @Override
//    public void register(StreamObserver<Response> responseObserver) {
//        LOGGER.info("adding observer " + responseObserver);
//        observers.add(responseObserver);
//        return new StreamObserver<Request>() {
//            @Override
//            public void onNext(Request value) {
//
//            }
//
//            @Override
//            public void onError(Throwable t) {
//
//            }
//
//            @Override
//            public void onCompleted() {
//                LOGGER.info("removing observer " + responseObserver);
//                responseObserver.onCompleted();
//                observers.remove(responseObserver);
//            }
//        };
//    }

